File: r/R/harmonize.R
#' Harmonize an EAVS year to stable concept names
#'
#' Renames a single survey year's raw variables to the concept names shared
#' across years, so that different years line up. For example, mail ballots
#' rejected is `C4a` in 2020 but `C9a` in 2024; both become `mail_rejected`.
#' This is the step that makes years comparable, and it is what turns EAVS's
#' silent renumbering from a trap into a non-issue.
#'
#' Only the curated concepts in [eavs_dictionary] are renamed. By default other
#' columns are dropped; set `keep_unmatched = TRUE` to keep them under their
#' original names.
#'
#' @param data A single survey year of EAVS data, typically from
#'   [eavs_recode_missing()] (so item columns are numeric). A `year` column, as
#'   added by [eavs_read()], is used to pick the year.
#' @param year The survey year, if `data` has no `year` column.
#' @param keep_unmatched Keep columns that are not in the dictionary, under
#'   their original names. Defaults to `FALSE`.
#' @param dictionary The dictionary to harmonize against. Defaults to the
#'   bundled [eavs_dictionary]; pass your own to use an extended crosswalk.
#'
#' @return A tibble with identifier and concept columns named consistently
#'   across years, plus `year`.
#' @seealso [eavs_load()] to download, recode, and harmonize in one step.
#' @export
eavs_harmonize <- function(data, year = NULL, keep_unmatched = FALSE,
                           dictionary = NULL) {
  year <- resolve_year(data, year)
  dict_y <- dict_for_year(year, dictionary)

  matched <- dict_y[dict_y$code %in% names(data), c("concept", "code", "section")]
  if (nrow(matched) == 0) {
    cli::cli_abort(c(
      "None of {.arg data}'s columns match the {year} dictionary.",
      i = "Is {.arg data} really the {year} EAVS?"
    ))
  }

  out <- data
  idx <- match(matched$code, names(out))
  names(out)[idx] <- matched$concept

  lead <- intersect(c("year", "survey"), names(out))
  ids <- matched$concept[matched$section == "id"]
  items <- matched$concept[matched$section != "id"]
  keep <- c(lead, ids, items)
  if (keep_unmatched) {
    keep <- c(keep, setdiff(names(out), keep))
  }
  out <- out[, keep, drop = FALSE]

  if (!"year" %in% names(out)) {
    out <- tibble::add_column(out, year = year, .before = 1)
  }
  tibble::as_tibble(out)
}

File: py/src/tidyeavs/harmonize.py
"""Renaming a survey year's raw codes to stable concept names."""

from __future__ import annotations

import pandas as pd

from ._dictionary import dict_for_year, resolve_year


def harmonize(
    data: pd.DataFrame,
    year: int | None = None,
    keep_unmatched: bool = False,
    dictionary: pd.DataFrame | None = None,
) -> pd.DataFrame:
    """Rename one year's raw variables to the concept names shared across years.

    Mail ballots rejected is ``C4a`` in 2020 and ``C9a`` in 2024; both become
    ``mail_rejected``. This is the step that makes years comparable, and it is
    what turns EAVS's renumbering between cycles from a trap into a non-issue.

    Only the curated concepts in the crosswalk are renamed. Other columns are
    dropped unless ``keep_unmatched=True``, which keeps them under their
    original names.

    ``data`` is typically the output of :func:`tidyeavs.recode_missing`, so item
    columns are already numeric. The year is taken from a ``year`` column when
    there is one, as :func:`tidyeavs.read` adds.
    """
    year = resolve_year(data, year)
    dict_year = dict_for_year(year, dictionary)

    matched = dict_year[dict_year["code"].isin(data.columns)]
    if matched.empty:
        raise ValueError(
            f"None of the columns match the {year} dictionary. "
            f"Is this really the {year} EAVS?"
        )

    renames = dict(zip(matched["code"], matched["concept"]))
    out = data.rename(columns=renames)

    lead = [name for name in ("year", "survey") if name in out.columns]
    ids = [c for c in matched.loc[matched["section"] == "id", "concept"]]
    concepts = [c for c in matched.loc[matched["section"] != "id", "concept"]]

    keep = lead + ids + concepts
    if keep_unmatched:
        keep += [name for name in out.columns if name not in keep]

    out = out[keep]
    if "year" not in out.columns:
        out = out.copy()
        out.insert(0, "year", year)
    return out.reset_index(drop=True)
